policykit/policyengine/models.py

```diff
@@ -375,179 +367,156 @@ class Proposal(models.Model):
-    )
-    """The datastore containing any additional data for the action. May or may not exist."""
-
-    class Meta:
-        abstract = True
-
-class ConstitutionAction(BaseAction, PolymorphicModel):
-    """Constitution Action"""
-
-    # NOTE: Why is this duplicated here?
-    community = models.ForeignKey(CommunityPlatform, models.CASCADE)
-
-    # NOTE: Should be moved to BaseAction
-    initiator = models.ForeignKey(CommunityUser, models.CASCADE, null=True)
-    """The User who initiated the action. May not exist if initiated by PolicyKit."""
-
-    # NOTE: Why is this duplicated here?
-    is_bundled = models.BooleanField(default=False)
-
-    # NOTE: Should be moved to BaseAction
-    action_type = "ConstitutionAction"
-    """Type of action (Constitution or Platform)."""
+    action_kind = None # platform vs constitution. don't override
 
-    # FIXME: Move this to BaseAction
-    action_codename = ''
-    """The codename of the action."""
-
-    class Meta:
-        verbose_name = 'constitutionaction'
-        verbose_name_plural = 'constitutionactions'
-
-    def pass_action(self):
-        """
-        Sets the action's proposal to PASSED.
+    def save(self, *args, **kwargs):
         """
-        self.proposal.status = Proposal.PASSED
-        self.proposal.save()
-        actstream_action.send(self, verb='was passed', community_id=self.community.id, action_codename=self.action_codename)
+        Saves the action. If new, evaluates against current policies. Note: Only meant for internal use.
 
-    def fail_action(self):
+        :meta private:
         """
-        Sets the action's proposal to FAILED.
-        """
-        self.proposal.status = Proposal.FAILED
-        self.proposal.save()
-        actstream_action.send(self, verb='was failed', community_id=self.community.id, action_codename=self.action_codename)
+        if not self.pk:
+            # Runs if initiator has propose permission, OR if there is no initiator.
+            can_propose_perm = f"{self._meta.app_label}.add_{self.action_type}"
+            if not self.initiator or self.initiator.has_perm(can_propose_perm):
+                super(BaseAction, self).save(*args, **kwargs)
+                engine.govern_action(self)
 
-    def shouldCreate(self):
-        """
-        True if the action needs to be created. Note: Only meant for internal use.
-        """
-        return not self.pk # Runs only when object is new
+        super(BaseAction, self).save(*args, **kwargs)
 
-    def save(self, *args, **kwargs):
-        """
-        Saves the action. If new, checks it against current passed policies. Note: Only meant for internal use.
+    @property
+    def action_type(self):
+        """The type of action (such as 'slackpostmessage' or 'policykitaddcommunitydoc')."""
+        return self._meta.model_name
+
+    @property
+    def action_codename(self):
+        """Same as ``action_type``. Added for backwards-compatibility.
+        :meta private:
         """
```

Comment:
I removed the `action_codename` and `action_type` static attributes in favor of this approach. The codenames *must* be the same as the model name, because the PolicyKit engine relies on the Django-generated `add_<model name>` permissions to decide whether certain actions can be proposed.